### DISCUSSION TITLE
Improving image quality on IMX477

### DISCUSSION THREAD

**Discussion by caphector** [aaronwmorris/indi-allsky#930] [Q&A]
I set up indi-allsky to run a camera I set up for a few different reasons, including using it as a weather cam, plane-spotting, and astronomical observations. As part of this I'd like to get naturalistic images during the day, but even after tweaking settings I'm still getting color casts and I wanted to see if I was missing a better way to improve color fidelity.

I have an IMX477 camera; I currently have it configured in DNG mode with AWB enabled and R/G/B 3.22/1.0/1.5 and Bayer set manually at BGGR (FWIW, changing from Auto Detect to BGGR improved visual quality enough that I started tweaking the balance factors). This is an example of what I see from indi-allsky on the left and from raspistill on the right:

![compare](https://github.com/aaronwmorris/indi-allsky/assets/8420750/924e62c2-ef09-419c-a6e4-5cba5fb63512)

These were taken at about the same time. Whenever I look at the levels in the indi-allsky image they're oddly peaky as well:

![image](https://github.com/aaronwmorris/indi-allsky/assets/8420750/9fa5e8bb-1631-477a-a3ad-cd67e3e068e3)

**Comment by aaronwmorris** [comment]
Does your camera have the IR/cut filter?

**Reply by caphector** [reply]
It's the stock Pi HQ camera; I didn't remove the IR filter so yes it does.

**Reply by aaronwmorris** [reply]
If you enable AWB with DNG raw mode, your R and B color adjustment values will actually be ignored and values from libcamera are used instead.

If you use AWB with JPEG mode, you should see the same results as raspistill.  This is because libcamera performs a bunch of image adjustments before producing the final image.

With DNG and AWB, you get the RAW values from the image sensor, but the image adjustments are not applied.  The image adjustments it would have made are included in a JSON metadata output, and these have to be applied manually in software.  These values are in the form of color gain adjustments for the R and B channels, however, I have not been able to find the actual mathematical formula used to apply the color gain values.  I am using straight multiplication, which is a close result, but not quite right.

It is very difficult to produce the same results.

**Reply by caphector** [reply]
I'd thought that tweaking the RGB settings under Config -> Image would be used in addition to the AWB setting (`Auto WB is applied after manual WB settings`). After disabling AWB my sky looks more natural.

I prefer DNG because it seems to have better quality overall at night, even as I still need to tweak a bit. Is there anything I can do to reduce clipping in extremely bright areas during the day? 

I moved my camera to the roof last night and I'm now seeing a lot of skyglow in the corners (Bottle 6/7 around here). Is the Camera -> Night Gain setting the right thing to adjust? I didn't see much change when I was tweaking the night ADU.

**Reply by aaronwmorris** [reply]
Keep in mind there are two different AWB settings.  One for libcamera on the Camera tab and the software AWB that is on the Image tab.

The libcamera AWB settings would be applied as the manual AWB step since those are the values used for the color gain adjustments.  The software AWB just tries to somewhat align the average of the R, G, and B channels, but is not always great.

This page has the order of operations for the whole workflow:  https://github.com/aaronwmorris/indi-allsky/wiki/Image-order-of-operations

DNG is definitely the way to go.  It also allows you to take proper dark frames.

You have the correct settings for gain and ADU.  You should also consider the ADU Field of View [FoV] and possibly the detection mask (Processing tab).  I have a wiki page on detection masks:  https://github.com/aaronwmorris/indi-allsky/wiki/Detection-Masks

**Reply by caphector** [reply]
I am currently only using the Image tab white balance settings; all of the AWB settings are disabled. 3.22 / 1.0 / 1.5 on the Image tab seems to work very well for this camera.

I've added a detection mask and I'll keep tweaking the night gain settings. Is some degree of exposure banding / flapping expected? I notice it particularly in banding on keograms.

**Reply by aaronwmorris** [reply]
If you had some examples of what you are seeing.  Exposure flapping does happen, but usually not often.  There are some strategies to helping with it.

Increasing the `Target ADU Deviation` can help reduce flapping (assuming we are talking about the same kind of flapping).

**Reply by caphector** [reply]
Here's a timelapse and keogram showing the exposure banding in different ways: 

https://www.caphector.com/allsky/videos/allsky-timelapse_ccd1_20230908_day.mp4
https://www.caphector.com/allsky/keograms/allsky-keogram_ccd1_20230907_day.jpg

**Reply by aaronwmorris** [reply]
The exposure banding ("stepping" might be a better word) as you called it is pretty normal for the morning and evening when there are drastic changes in lighting.  It will crop up as the sun or moon moves across the view, because they skew the ADU values.

Exposure flapping is a different phenomenon.  When the exposure algorithm needs to calculate a new exposure, sometimes it will get stuck between two exposures trying to find a stable value and "flap" back and forth which results in a strobe effect.  This usually only happens during the day, at extremely low exposure values.

**Reply by caphector** [reply]
Is there anything I can do to pull more detail out of bright areas? I've noticed that in bright area detail gets bloomed out and I wonder if more detail can be recovered.

I've been slowly lowering my gain, but since the camera is shooting DNG would it be possible to use a image direct off the sensor to test different image settings? While I don't mind the current process it would be nice to compare settings changes against the same image rather than checking after the fact.

**Reply by aaronwmorris** [reply]
Testing settings with a RAW file is one of the things I would like to add soon.

There is probably not much that can be done with bright areas.  When you have things like clouds pass by and their brightness is too high, they just end up being overexposed.  If you increase/decrease the exposure, the clouds will be normalized and the rest of the image will be too dark.

You can try the stretching option, but I doubt it will do anything to fix areas being too bright.  CLAHE contrast enhance might help a little, but that can have some weird side effects.

You have some pretty gnarly lights that are shining on the dome that is killing your contrast.

**Reply by caphector** [reply]
Yeah; I need to get the camera on the side of the roof away from the streetlights, but that requires getting power up there as well. I'll put that on the To Do list. I'd hoped that painting the edges of the dome black to cut down on glare would help, but no dice.

**Reply by caphector** [reply]
Got it moved to the dark side of the roof. Helpfully, the moon was nearly full when I moved it but I'm looking forward to seeing star trails getting generated again once it wanes a bit.

I'll need to start poking the gain settings again; would be nice to have the sky center closer to black to pull the corners brightness down further.

**Comment by caphector** [comment]
Thanks for the explanation and for confirming that the behavior is normal; the visual look of the exposure change in the timelapse in particular made me wonder if it was flapping.
